File: Satellite.cpp
#include "Satellite.h"
#include <cstring>
#include <iostream>
#include <queue>
#include "Constellation.h"
#include "utils.h"
using namespace std;
void Satellite::resetRoutingTable() {
    if (cons->routingStrategy == 1 || cons->routingStrategy == 3 || cons->routingStrategy == 6) {
        canVis.clear();
        passTo.clear();
        latency.clear();
        for (int i = 0; i < cons->nOrbit * cons->nSatPerOrbit; i++) {
            canVis.push_back(false);
            passTo.push_back(-1);
            latency.push_back(-1);
        }
    }
    if (cons->routingStrategy == 3) {
        passTo3.clear();
        passTo3Calced.clear();
        for (int i = 0; i < cons->nOrbit * cons->nSatPerOrbit; i++) {
            vector<int> tmp;
            passTo3.push_back(tmp);
            passTo3Calced.push_back(false);
        }
    }
}
void Satellite::resetISL() {
    islCnt = 0;
    islAvailability.clear();
    islTarget.clear();
}
void Satellite::enableAllISL() {
    for (int i = 0; i < islCnt; i++) {
        islAvailability[i] = true;
    }
}
void Satellite::addISL(Satellite* target) {
    for (auto sat = islTarget.begin(); sat != islTarget.end(); sat++) {
        if (*sat == target) {
            cout << "Error: " + this->str() + " already had ISL with " + target->str() + ". " << endl;
            return;
        }
    }
    islCnt++;
    islAvailability.push_back(true);
    islTarget.push_back(target);
}
void Satellite::disableISL(Satellite* target) {
    for (int i = 0; i < islCnt; i++) {
        if (islTarget[i] == target) {
            islAvailability[i] = false;
            return;
        }
    }
    cout << "Error: " + this->str() + " has no ISL with " + target->str() + ". " << endl;
}
void Satellite::enableISL(Satellite* target) {
    for (int i = 0; i < islCnt; i++) {
        if (islTarget[i] == target) {
            islAvailability[i] = true;
            return;
        }
    }
    cout << "Error: " + this->str() + " has no ISL with " + target->str() + ". " << endl;
}
Satellite* Satellite::shiftSat(int orbitShift, int phaseShift) {
    int targetOrbit = curOrbit + orbitShift;
    int targetPosInOrbit = curPosInOrbit + phaseShift;
    int nOrbit = cons->nOrbit;
    int nSatPerOrbit = cons->nSatPerOrbit;
    targetOrbit = ((targetOrbit - 1) % nOrbit + nOrbit) % nOrbit + 1;
    targetPosInOrbit = ((targetPosInOrbit - 1) % nSatPerOrbit + nSatPerOrbit) % nSatPerOrbit + 1;
    return cons->getSat(targetOrbit, targetPosInOrbit);
}
void Satellite::buildRoutingTable(double t) {
    resetRoutingTable();
    if (cons->routingStrategy == 1 || cons->routingStrategy == 3) {
        buildRoutingTable1(this, t);
    }
    if (cons->routingStrategy == 6) {
        buildRoutingTable6(this, t);
    }
}

void buildRoutingTable1(Satellite* sat, double t) {
    priority_queue<Node> q;
    sat->resetRoutingTable();
    sat->latency[sat->id] = 0;
    q.push(Node(sat, 0));

    sat->canVis[sat->id] = true;
    sat->passTo[sat->id] = -1;
    Satellite* curSat = sat;
    for (int i = 0; i < sat->islCnt; i++) {
        Satellite* targetSat = curSat->islTarget[i];
        // cout << targetSat->id << " " << targetSat->str() << endl
        //      << curSat->id << " " << curSat->str() << endl;
        if ((sat->latency[targetSat->id] == -1) ||
            sat->latency[curSat->id] + satLatency(curSat, targetSat, t) < sat->latency[targetSat->id]) {
            sat->latency[targetSat->id] = sat->latency[curSat->id] + satLatency(curSat, targetSat, t);
            sat->passTo[targetSat->id] = i;
            q.push(Node(targetSat, sat->latency[targetSat->id]));
        }
    }

    while (!q.empty()) {
        Satellite* curSat = q.top().sat;
        q.pop();
        if (sat->canVis[curSat->id]) {
            continue;
        }
        sat->canVis[curSat->id] = true;
        for (int i = 0; i < curSat->islCnt; i++) {
            Satellite* targetSat = curSat->islTarget[i];
            if ((sat->latency[targetSat->id] == -1) ||
                sat->latency[curSat->id] + satLatency(curSat, targetSat, t) < sat->latency[targetSat->id]) {
                sat->latency[targetSat->id] = sat->latency[curSat->id] + satLatency(curSat, targetSat, t);
                sat->passTo[targetSat->id] = sat->passTo[curSat->id];
                q.push(Node(targetSat, sat->latency[targetSat->id]));
            }
        }
    }
}

void buildRoutingTable6(Satellite* sat, double t) {
    priority_queue<Node> q;
    sat->resetRoutingTable();
    sat->latency[sat->id] = 0;
    q.push(Node(sat, 0));

    sat->canVis[sat->id] = true;
    sat->passTo[sat->id] = -1;
    Satellite* curSat = sat;
    for (int i = 0; i < sat->islCnt; i++) {
        Satellite* targetSat = curSat->islTarget[i];
        // cout << targetSat->id << " " << targetSat->str() << endl
        //      << curSat->id << " " << curSat->str() << endl;
        if ((sat->latency[targetSat->id] == -1) || sat->latency[curSat->id] + 0.001 < sat->latency[targetSat->id]) {
            sat->latency[targetSat->id] = sat->latency[curSat->id] + 0.001;
            sat->passTo[targetSat->id] = i;
            q.push(Node(targetSat, sat->latency[targetSat->id]));
        }
    }

    while (!q.empty()) {
        Satellite* curSat = q.top().sat;
        q.pop();
        if (sat->canVis[curSat->id]) {
            continue;
        }
        sat->canVis[curSat->id] = true;
        for (int i = 0; i < curSat->islCnt; i++) {
            Satellite* targetSat = curSat->islTarget[i];
            if ((sat->latency[targetSat->id] == -1) || sat->latency[curSat->id] + 0.001 < sat->latency[targetSat->id]) {
                sat->latency[targetSat->id] = sat->latency[curSat->id] + 0.001;
                sat->passTo[targetSat->id] = sat->passTo[curSat->id];
                q.push(Node(targetSat, sat->latency[targetSat->id]));
            }
        }
    }
}

void Satellite::buildRoutingTable3(Satellite* target, double t) {
    buildRoutingTable3Help(this, target, t);
}

void buildRoutingTable3Help(Satellite* sat, Satellite* target, double t) {
    if (sat->passTo3Calced[target->id]) {
        return;
    }
    sat->passTo3Calced[target->id] = true;

    Constellation* cons = sat->cons;
    bool used[1700][6] = {0};
    Satellite* curSat = sat;
    while (curSat != target && target != sat) {
        used[curSat->id][curSat->passTo[target->id]] = true;
        curSat = curSat->islTarget[curSat->passTo[target->id]];
    }
    priority_queue<Node3> q;
    double latency[1700] = {0};
    bool vis[1700] = {0};
    for (int i = 0; i < cons->nOrbit * cons->nSatPerOrbit; i++) {
        latency[i] = -1;
    }
    latency[sat->id] = 0;
    vector<int> tmp;
    q.push(Node3(sat, 0, tmp));
    while (!q.empty()) {
        Node3 curNode = q.top();
        q.pop();
        Satellite* curSat = curNode.sat;
        if (vis[curSat->id]) {
            continue;
        }
        vis[curSat->id] = true;
        if (curSat == target) {
            sat->passTo3[target->id] = curNode.pathRecord;
            break;
        }
        for (int i = 0; i < curSat->islCnt; i++) {
            if (used[curSat->id][i]) {
                continue;
            }
            Satellite* targetSat = curSat->islTarget[i];
            if ((latency[targetSat->id] == -1) ||
                latency[curSat->id] + satLatency(curSat, targetSat, t) < latency[targetSat->id]) {
                latency[targetSat->id] = latency[curSat->id] + satLatency(curSat, targetSat, t);
                vector<int> newPathRecord = curNode.pathRecord;
                newPathRecord.push_back(i);
                q.push(Node3(targetSat, latency[targetSat->id], newPathRecord));
            }
        }
    }
}

Point Satellite::pos(double t) {
    return calcPos(this, t).check();
}
xyzPoint Satellite::xyzPos(double t) {
    return ang2xyz(pos(t).check());
}
void Node::copyFrom(const Node& target) {
    sat = target.sat;
    dis = target.dis;
}
Node::Node(const Node& target) {
    copyFrom(target);
}
Node& Node::operator=(const Node& target) {
    copyFrom(target);
    return *this;
}
bool operator<(const Node& a, const Node& b) {
    return a.dis > b.dis;
}
void Node3::copyFrom(const Node3& target) {
    sat = target.sat;
    dis = target.dis;
    pathRecord = target.pathRecord;
}
Node3::Node3(const Node3& target) {
    copyFrom(target);
}
Node3& Node3::operator=(const Node3& target) {
    copyFrom(target);
    return *this;
}
bool operator<(const Node3& a, const Node3& b) {
    return a.dis > b.dis;
}
void Satellite::printISL() {
    cout << str() << " is connected with:" << endl;
    for (int i = 0; i < islCnt; i++) {
        cout << "    " << (islAvailability[i] ? "(on) " : "(off)") << "  " << islTarget[i]->str() << endl;
    }
}
void Satellite::printISLandPos(double t) {
    printPos(t);
    printISL();
}
void Satellite::printPos(double t) {
    cout << str() << " is at " << pos(t).str() << endl;
}
void Satellite::printRoutingTable(double t) {
    cout << str() << "'s routing table:" << endl;
    if (cons->routingStrategy == 1) {
        for (int i = 0; i < cons->nOrbit * cons->nSatPerOrbit; i++) {
            cout << "  To " << cons->sats[i]->str() << (canVis[i] ? " (can vis): pass to " : " (not vis)") << " "
                 << (canVis[i] ? islTarget[pass(cons->sats[i], t)]->str() : "") << endl;
        }
    }
    if (cons->routingStrategy == 3) {
        for (int i = 0; i < cons->nOrbit * cons->nSatPerOrbit; i++) {
            cout << "  To " << cons->sats[i]->str() << ": " << endl;
            Satellite* curSat = this;
            cout << "    ";
            while (curSat->id != i) {
                cout << curSat->passTo[i] << " ";
                curSat = curSat->islTarget[curSat->passTo[i]];
            }
            cout << endl;
            if (passTo3Calced[i]) {
                cout << "    ";
                curSat = this;
                for (auto islId = passTo3[i].begin(); islId != passTo3[i].end(); islId++) {
                    cout << *islId << " ";
                    curSat = curSat->islTarget[*islId];
                }
                cout << endl;
            }
        }
    }
}
int Satellite::pass(Satellite* target, double t) {
    return passTo[target->id];
}
int Satellite::pass(Point target, double t) {
    xyzPoint targetXyz = ang2xyz(target);
    double bestLatency = xyzPointLatency(targetXyz, xyzPos(t));
    int bestISL = -1;
    for (int i = 0; i < islCnt; i++) {
        if (!islAvailability[i]) {
            continue;
        }
        double iLatency = xyzPointLatency(targetXyz, islTarget[i]->xyzPos(t));
        if (iLatency < bestLatency) {
            bestLatency = iLatency;
            bestISL = i;
        }
    }
    return bestISL;
}
int Satellite::passByGrid(Satellite* target, double t) {
    double bestDist = gridDist(this, target, t);
    int bestISL = -1;
    for (int i = 0; i < islCnt; i++) {
        if (!islAvailability[i]) {
            continue;
        }
        double dist = gridDist(islTarget[i], target, t);
        if (dist < bestDist) {
            bestDist = dist;
            bestISL = i;
        }
    }
    return bestISL;
}

File: models/Satellite.py
from .Point import Point, xyzPoint
import sys
from os.path import dirname, abspath
import heapq

if dirname(dirname(abspath(__file__))) not in sys.path:
    sys.path.append(dirname(dirname(abspath(__file__))))

from utils import calcPos, ang2xyz, calcIndex, satLatency


class Satellite:
    satelliteCnt = 0

    def __init__(self,  cons: 'Constellation', consId: int, curOrbit: int, curPosInOrbit: int) -> None:
        Satellite.satelliteCnt += 1
        self.id = Satellite.satelliteCnt
        self.cons = cons
        self.consId = consId
        self.available = True
        self.index = calcIndex(curOrbit, curPosInOrbit, cons)
        # print(consId)

        # ============= INITIAL POSITION ============
        self.curOrbit = curOrbit
        self.curPosInOrbit = curPosInOrbit

        # ============= INITIAL ISL ==============
        self.islCnt = 0
        self.islAvailability = []
        self.islTarget = []

        # ============= INITIAL ROUTING TABLE ================
        self.canVis = []
        # if self.canVis[i]=false, self.passTo[i] is undefined
        # i: index of sat in cons
        # passTo stores satellite object
        self.passTo = []
        self.latency = []

    def __str__(self) -> str:
        return f'satellite #{self.id} at ({self.curOrbit},{self.curPosInOrbit}) ({"on" if self.available else "off"})'

    def pos(self, t: float) -> Point:
        # return the alt lon lat position of a sat
        return calcPos(self, t).check()

    def xyzPos(self, t: float) -> xyzPoint:
        # return the xyz position of a sat
        return ang2xyz(calcPos(self, t).check())

    def addISL(self, target: 'Satellite') -> None:
        # add ISL with target
        try:
            self.islTarget.index(target)
        except:
            self.islCnt += 1
            self.islAvailability.append(True)
            self.islTarget.append(target)
        else:
            print(f'Error: {self} already had ISL with {target}. ')

    def resetISL(self) -> None:
        # reset ISL of a sat
        self.islCnt = 0
        self.islAvailability.clear()
        self.islTarget.clear()

    def removeISL(self, target: 'Satellite') -> None:
        # remove ISL with target
        if (self.islCnt == 0):
            print(
                f'Error: {self} has 0 ISL. Cannot remove {target} from ISL connection.')
            return
        try:
            index = self.islTarget.index(target)
            self.islTarget.remove(target)
            self.islAvailability.pop(index)
            self.islCnt -= 1
        except:
            print(f'Error: {self} has no ISL with {target}.')

    def disableISL(self, target: 'Satellite') -> None:
        # disable ISL with target
        if (self.islCnt == 0):
            print(
                f'Error: {self} has 0 ISL. Cannot disable {target} from ISL connection.')
            return
        try:
            index = self.islTarget.index(target)
            self.islAvailability[index] = False
        except:
            print(f'Error: {self} has no ISL with {target}.')

    def enableISL(self, target: 'Satellite') -> None:
        # enable ISL with target
        if (self.islCnt == 0):
            print(
                f'Error: {self} has 0 ISL. Cannot enable {target} from ISL connection.')
            return
        try:
            index = self.islTarget.index(target)
            self.islAvailability[index] = True
        except:
            print(f'Error: {self} has no ISL with {target}.')

    def enableAllISL(self) -> None:
        # enable all ISL in the satellite
        for i in range(0, self.islCnt):
            self.islAvailability[i] = True

    def shiftSat(self, orbitShift: int, phaseShift: int) -> 'Satellite':
        # calculate the satellite shifted by (orbitShift, phaseShift)
        # e.g. (1,1).shiftSat(-1,1)=(24,2), if the constellation has 24 orbits
        targetOrbit = self.curOrbit+orbitShift
        targetPosInOrbit = self.curPosInOrbit+phaseShift
        nOrbit = self.cons.nOrbit
        nSatPerOrbit = self.cons.nSatPerOrbit
        targetOrbit = ((targetOrbit-1) % nOrbit+nOrbit) % nOrbit+1
        targetPosInOrbit = ((targetPosInOrbit-1) %
                            nSatPerOrbit+nSatPerOrbit) % nSatPerOrbit+1
        return self.cons.getSat(targetOrbit, targetPosInOrbit)

    def printISL(self) -> None:
        # print the ISLs
        print(f'{self} is connected with:')
        for i in range(0, self.islCnt):
            print(
                f'    {"(on) " if self.islAvailability[i] else "(off)"}  {self.islTarget[i]}')

    def printISLandPos(self, t: float) -> None:
        # print ISLs and pos
        print(f'{self} is at {self.pos(t)}')
        self.printISL()

    def resetRoutingTable(self) -> None:
        # reset routing table of a sat
        self.canVis = [False for i in range(
            0, self.cons.nOrbit*self.cons.nSatPerOrbit+1)]
        self.passTo = [i for i in range(
            0, self.cons.nOrbit*self.cons.nSatPerOrbit+1)]
        self.latency = [-1 for i in range(0,
                                          self.cons.nOrbit*self.cons.nSatPerOrbit+1)]

    def buildRoutingTable(self, t) -> None:
        # build routing table for the satellite
        self.resetRoutingTable()
        cons = self.cons
        if (cons.routingStrategy == 1 or cons.routingStrategy == 2):
            buildRoutingTable1(self, t)

    def printRoutingTable(self) -> None:
        # print the routing table
        print(f"{self}'s routing table: ")
        for i in range(0, self.cons.nOrbit*self.cons.nSatPerOrbit):
            print(
                f'    To {self.cons.sats[i]} {"(can vis): pass to " if self.canVis[i] else "(not vis)"}{self.passTo[i] if self.canVis[i] else ""} {self.latency[i] if self.canVis[i] else ""}')


class HeapQNode:
    # the node class of heap queue in Dijkstra's algo
    def __init__(self, sat, latency):
        self.sat = sat
        self.latency = latency

    def __lt__(self, other: 'HeapQNode'):
        return self.latency < other.latency

    def __eq__(self, other: 'HeapQNode'):
        return self.latency == other.latency

    def __repr__(self) -> str:
        return f'({self.sat}: {self.latency})'


def buildRoutingTable1(sat: 'Satellite', t: float) -> None:
    # build routing table with strategy 1
    # print(f'building routing table for {sat}')
    hq = []
    cons = sat.cons
    heapq.heapify(hq)
    heapq.heappush(hq, HeapQNode(sat, 0))
    sat.latency = [-1 for i in range(0, cons.nOrbit*cons.nSatPerOrbit+1)]
    sat.latency[sat.index] = 0

    curSat = heapq.heappop(hq).sat
    sat.canVis[curSat.index] = True
    sat.passTo[curSat.index] = curSat
    for i in range(0, curSat.islCnt):
        targetSat = curSat.islTarget[i]
        islAvailability = curSat.islAvailability[i]
        if ((sat.latency[targetSat.index] == -1) or sat.latency[curSat.index]+satLatency(curSat, targetSat, t) < sat.latency[targetSat.index]):
            sat.latency[targetSat.index] = sat.latency[curSat.index] + \
                satLatency(curSat, targetSat, t)
            sat.passTo[targetSat.index] = targetSat
            heapq.heappush(hq, HeapQNode(
                targetSat, sat.latency[targetSat.index]))
            # print(f'    {targetSat} {sat.latency[targetSat.index]}')

    while (len(hq) > 0):
        # print(f'{curSat}: {sat.latency[curSat.index]}')
        # print(hq)
        curSat = heapq.heappop(hq).sat
        if (sat.canVis[curSat.index]):
            continue
        sat.canVis[curSat.index] = True
        for i in range(0, curSat.islCnt):
            targetSat = curSat.islTarget[i]
            islAvailability = curSat.islAvailability[i]
            if ((sat.latency[targetSat.index] == -1) or sat.latency[curSat.index]+satLatency(curSat, targetSat, t) < sat.latency[targetSat.index]):
                sat.latency[targetSat.index] = sat.latency[curSat.index] + \
                    satLatency(curSat, targetSat, t)
                sat.passTo[targetSat.index] = sat.passTo[curSat.index]
                heapq.heappush(hq, HeapQNode(
                    targetSat, sat.latency[targetSat.index]))
                # print(f'    {targetSat} {sat.latency[targetSat.index]}')
